akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
+              openRequests = openRequests.enqueue(r)
+              messageEndPending = r.createEntity.isInstanceOf[StreamedEntityCreator[_, _]]
+              val rs = if (r.expect100Continue) {
+                oneHundredContinueResponsePending = true
+                r.copy(createEntity = with100ContinueTrigger(r.createEntity))
+              } else r
+              push(requestPrepOut, rs)
+            case MessageEnd ⇒
+              messageEndPending = false
+              push(requestPrepOut, MessageEnd)
+            case MessageStartError(status, info) ⇒ finishWithIllegalRequestError(status, info)
+            case x                               ⇒ push(requestPrepOut, x)
           }
+        override def onUpstreamFinish() =
+          if (openRequests.isEmpty) completeStage()
+          else complete(requestPrepOut)
       })
 
-      setHandler(applicationInput, new InHandler {
-        override def onPush(): Unit = {
-          val response = grab(applicationInput)
-          // see the comment on [[OneHundredContinue]] for an explanation of the closing logic here (and more)
-          val close = requestStart.closeRequested || requestStart.expect100ContinueResponsePending
-          abortReading(oneHundredContinue)
-          emit(out, ResponseRenderingContext(response, requestStart.method, requestStart.protocol, close),
-            if (close) () ⇒ completeStage() else pullBypass)
+      setHandler(requestPrepOut, new OutHandler {
+        def onPull(): Unit =
+          if (oneHundredContinueResponsePending) pullSuppressed = true
+          else pull(requestParsingIn)
+        override def onDownstreamFinish() = cancel(requestParsingIn)
+      })
+
+      setHandler(httpResponseIn, new InHandler {
+        def onPush(): Unit = {
+          val response = grab(httpResponseIn)
+          val requestStart = openRequests.head
+          openRequests = openRequests.tail
+          val isEarlyResponse = messageEndPending && openRequests.isEmpty
+          if (isEarlyResponse && response.status.isSuccess)
+            log.warning(
+              """Sending 2xx response before end of request was received...
+                |Note that the connection will be closed after this response. Also, many clients will not read early responses!
+                |Consider waiting for the request end before dispatching this response!""".stripMargin)
+          val close = requestStart.closeRequested ||
+            requestStart.expect100Continue && oneHundredContinueResponsePending ||
+            isClosed(requestParsingIn) && openRequests.isEmpty ||
+            isEarlyResponse
+          push(responseCtxOut, ResponseRenderingContext(response, requestStart.method, requestStart.protocol, close))
+          if (close) complete(responseCtxOut)
         }
+        override def onUpstreamFinish() =
+          if (openRequests.isEmpty && isClosed(requestParsingIn)) completeStage()
+          else complete(responseCtxOut)
```

Comment:
Because in certain cases (involving early responses!) the *entity* of the last response might depend on the entity of the last request still being read from the network. A simple example would be the classic "ping/pong"  where you stuff the request entity directly into the response. We actually have a test for exactly this setup.
If we complete the stage here we kill the incoming side as well and therefore the entity of the last response.